### PR TITLE
Remove deferral of Objects `SYNCED` events, and add sync event tests

### PR DESCRIPTION
**Pull request by lawrence-forooghian**
This does the following:

- removes the deferral of `SYNCED` events to the next event loop cycle (see commit message)
- adds tests for the `SYNCING` and `SYNCED` events based on the currently-implemented behaviour; will use this as a basis for https://github.com/ably/ably-liveobjects-swift-plugin/issues/80

## Summary by CodeRabbit

* **Refactor**
  * Improved object synchronization efficiency by streamlining internal state management, eliminating deferred event signaling in favor of immediate execution for enhanced responsiveness.

* **Tests**
  * Added comprehensive test suite for object synchronization events, validating sequences across multiple channel attachment scenarios and edge cases to ensure reliability.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: test/realtime/objects.test.js [hunk 1]
```diff
@@ -5518,5 +5518,203 @@ define(['ably', 'shared_helper', 'chai', 'objects', 'objects_helper'], function
...
+            { type: 'attached', hasObjects: true },
+            { type: 'objectSync', channelSerial: 'foo:1' },
+            { type: 'objectSync', channelSerial: 'foo:2' },
+            { type: 'objectSync', channelSerial: 'bar:1' },
+          ],
+          expectedSyncEvents: ['syncing'],
+        },
+      ];
+
+      forScenarios(this, syncEventsScenarios, async function (helper, scenario) {
```
Comment (VeskeR) [inline test/realtime/objects.test.js]: I'd suggest to always use `clientOptions` and `channelName` provided passed from `forScenarios` anyway. Client options are used for testing on protocols and transports and channel name includes description to make sure we don't accidentally share state between tests.
Even though not really relevant for these tests right now, it's a good idea to have it consistent in case we decide to change how tests work and need that functionality

File: test/realtime/objects.test.js [hunk 2]
```diff
@@ -5518,5 +5518,203 @@ define(['ably', 'shared_helper', 'chai', 'objects', 'objects_helper'], function
...
+
+        {
+          description:
+            'The first ATTACHED should always provoke a SYNCING even when HAS_OBJECTS is false, so that the SYNCED is preceded by SYNCING',
+          channelEvents: [{ type: 'attached', hasObjects: false }],
+          expectedSyncEvents: ['syncing', 'synced'],
+        },
+
+        {
+          description: 'ATTACHED with HAS_OBJECTS false once SYNCED should not provoke further events',
```
Comment (VeskeR) [inline test/realtime/objects.test.js]: Hmm, I wonder if this should be expected behavior.
When we receive HAS_OBJECTS false, it means that state is empty on a channel, and we will clear anything we currently have locally. This means that we do technically modify local data, just as if we have received HAS_OBJECTS=true that changed something.
And for subsequent HAS_OBJECTS=true we decided to emit `syncing -> synced` again, so maybe we should too for HAS_OBJECTS=false?

**Review by VeskeR — CHANGES_REQUESTED** [review]

File: test/realtime/objects.test.js (hunk 2 above)
Comment (lawrence-forooghian) [reply test/realtime/objects.test.js]: Sure, have added a commit for that. I don't really have opinions on it (as mentioned yesterday, I'm not sure it makes sense to emit SYNCING in _any_ of the cases where the sync data is updated in a one-shot fashion, but as we discussed we don't really have a great idea of how we intend for this API to be used, so agree that we should at least be consistent)

File: test/realtime/objects.test.js (hunk 1 above)
Comment (lawrence-forooghian) [reply test/realtime/objects.test.js]: Thanks for pointing that out, done.
